File: backend/ResultCollector.cs
/**
 * ResultCollector.cs - Host-side Result Collection Module
 *
 * ADB pull 실행 전 파일 존재 확인 및 동기화 로직
 *
 * Features:
 * 1. ADB shell을 통한 원격 파일 존재 확인
 * 2. .ready 신호 파일 확인 (Android 측 쓰기 완료 신호)
 * 3. 파일 크기 검증 후 pull 실행
 * 4. 재시도 로직 포함
 */

using System;
using System.Diagnostics;
using System.IO;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace DoaiDeviceFarm.Backend
{
    /// <summary>
    /// ADB를 통한 결과 파일 수집기
    /// </summary>
    public class ResultCollector
    {
        // 설정 상수
        private const int FILE_CHECK_TIMEOUT_MS = 30000;    // 파일 확인 최대 대기 시간
        private const int FILE_CHECK_INTERVAL_MS = 500;     // 파일 확인 간격
        private const int MIN_FILE_SIZE_BYTES = 1000;       // 최소 파일 크기 (1KB)
        private const int MAX_RETRY_COUNT = 3;              // 최대 재시도 횟수
        private const int RETRY_DELAY_MS = 1000;            // 재시도 간격

        // [RACE CONDITION FIX] 신호 감지 후 안전 대기 시간
        private const int SIGNAL_GRACE_PERIOD_MS = 500;     // job_finished 신호 감지 후 500ms 대기

        private readonly string _adbPath;
        private readonly string _localEvidencePath;

        public ResultCollector(string adbPath = "adb", string localEvidencePath = null)
        {
            _adbPath = adbPath;
            _localEvidencePath = localEvidencePath ?? Path.Combine(
                Environment.CurrentDirectory, "evidence"
            );

            // 로컬 증거 디렉토리 생성
            if (!Directory.Exists(_localEvidencePath))
            {
                Directory.CreateDirectory(_localEvidencePath);
            }
        }

        /// <summary>
        /// 원격 파일 존재 확인 (ADB shell)
        /// </summary>
        /// <param name="deviceSerial">기기 시리얼</param>
        /// <param name="remotePath">원격 파일 경로</param>
        /// <returns>파일 존재 여부</returns>
        public async Task<bool> CheckRemoteFileExistsAsync(string deviceSerial, string remotePath)
        {
            try
            {
                var result = await ExecuteAdbCommandAsync(
                    deviceSerial,
                    $"shell \"[ -f '{remotePath}' ] && echo EXISTS || echo NOTFOUND\""
                );

                return result.Trim() == "EXISTS";
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[ResultCollector] File check error: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// 원격 파일 크기 확인 (ADB shell)
        /// </summary>
        /// <param name="deviceSerial">기기 시리얼</param>
        /// <param name="remotePath">원격 파일 경로</param>
        /// <returns>파일 크기 (바이트), 실패 시 -1</returns>
        public async Task<long> GetRemoteFileSizeAsync(string deviceSerial, string remotePath)
        {
            try
            {
                var result = await ExecuteAdbCommandAsync(
                    deviceSerial,
                    $"shell stat -c %s '{remotePath}' 2>/dev/null || echo -1"
                );

                if (long.TryParse(result.Trim(), out long size))
                {
                    return size;
                }

                return -1;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[ResultCollector] Size check error: {ex.Message}");
                return -1;
            }
        }

        /// <summary>
        /// .ready 신호 파일 확인 및 파싱
        /// </summary>
        /// <param name="deviceSerial">기기 시리얼</param>
        /// <param name="remotePath">스크린샷 파일 경로</param>
        /// <returns>준비 상태 정보</returns>
        public async Task<ReadySignal> CheckReadySignalAsync(string deviceSerial, string remotePath)
        {
            var signalPath = remotePath + ".ready";

            try
            {
                // 신호 파일 존재 확인
                if (!await CheckRemoteFileExistsAsync(deviceSerial, signalPath))
                {
                    return new ReadySignal { Status = "not_found" };
                }

                // 신호 파일 내용 읽기
                var content = await ExecuteAdbCommandAsync(
                    deviceSerial,
                    $"shell cat '{signalPath}'"
                );

                if (string.IsNullOrWhiteSpace(content))
                {
                    return new ReadySignal { Status = "empty" };
                }

                // JSON 파싱
                var signal = JsonSerializer.Deserialize<ReadySignal>(content);
                return signal ?? new ReadySignal { Status = "parse_error" };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[ResultCollector] Signal check error: {ex.Message}");
                return new ReadySignal { Status = "error", Error = ex.Message };
            }
        }

        /// <summary>
        /// 파일 준비 완료까지 대기 (폴링)
        /// </summary>
        /// <param name="deviceSerial">기기 시리얼</param>
        /// <param name="remotePath">원격 파일 경로</param>
        /// <param name="timeoutMs">타임아웃 (밀리초)</param>
        /// <returns>준비 완료 여부</returns>
        public async Task<FileReadyResult> WaitForFileReadyAsync(
            string deviceSerial,
            string remotePath,
            int timeoutMs = FILE_CHECK_TIMEOUT_MS)
        {
            var result = new FileReadyResult
            {
                RemotePath = remotePath,
                StartTime = DateTime.UtcNow
            };

            var sw = Stopwatch.StartNew();

            Console.WriteLine($"[ResultCollector] Waiting for file: {remotePath}");

            while (sw.ElapsedMilliseconds < timeoutMs)
            {
                // 1. .ready 신호 파일 확인 (우선)
                var signal = await CheckReadySignalAsync(deviceSerial, remotePath);

                if (signal.Status == "ready")
                {
                    // [RACE CONDITION FIX] 신호 감지 후 500ms 안전 대기 시간
                    // Android 측 파일 시스템 버퍼가 완전히 플러시될 시간을 확보
                    Console.WriteLine($"[ResultCollector] Signal detected, waiting {SIGNAL_GRACE_PERIOD_MS}ms grace period...");
                    await Task.Delay(SIGNAL_GRACE_PERIOD_MS);

                    result.Success = true;
                    result.FileSize = signal.FileSize;
                    result.Duration = sw.Elapsed;
                    result.Method = "signal_file";

                    Console.WriteLine($"[ResultCollector] File ready (signal): {result.FileSize} bytes");
                    return result;
                }

                if (signal.Status == "failed")
                {
                    result.Success = false;
                    result.Error = signal.Error ?? "Remote capture failed";
                    result.Duration = sw.Elapsed;
                    result.Method = "signal_file";

                    Console.WriteLine($"[ResultCollector] File failed (signal): {result.Error}");
                    return result;
                }

                // 2. 직접 파일 확인 (신호 파일 없는 경우)
                if (await CheckRemoteFileExistsAsync(deviceSerial, remotePath))
                {
                    var size = await GetRemoteFileSizeAsync(deviceSerial, remotePath);

                    if (size >= MIN_FILE_SIZE_BYTES)
                    {
                        // 크기 안정화 확인 (1초 후 재확인)
                        await Task.Delay(1000);
                        var size2 = await GetRemoteFileSizeAsync(deviceSerial, remotePath);

                        if (size2 == size)
                        {
                            // [RACE CONDITION FIX] 직접 확인 시에도 500ms 안전 대기
                            Console.WriteLine($"[ResultCollector] Direct check stable, waiting {SIGNAL_GRACE_PERIOD_MS}ms grace period...");
                            await Task.Delay(SIGNAL_GRACE_PERIOD_MS);

                            result.Success = true;
                            result.FileSize = size;
                            result.Duration = sw.Elapsed;
                            result.Method = "direct_check";

                            Console.WriteLine($"[ResultCollector] File ready (direct): {result.FileSize} bytes");
                            return result;
                        }
                    }
                }

                await Task.Delay(FILE_CHECK_INTERVAL_MS);
            }

            // 타임아웃
            result.Success = false;
            result.Error = "Timeout waiting for file";
            result.Duration = sw.Elapsed;

            Console.WriteLine($"[ResultCollector] File timeout: {remotePath}");
            return result;
        }

        /// <summary>
        /// 안전한 ADB Pull (파일 준비 확인 후 실행)
        /// </summary>
        /// <param name="deviceSerial">기기 시리얼</param>
        /// <param name="remotePath">원격 파일 경로</param>
        /// <param name="localPath">로컬 저장 경로 (null이면 자동 생성)</param>
        /// <returns>Pull 결과</returns>
        public async Task<PullResult> SafePullAsync(
            string deviceSerial,
            string remotePath,
            string localPath = null)
        {
            var result = new PullResult
            {
                DeviceSerial = deviceSerial,
                RemotePath = remotePath,
                StartTime = DateTime.UtcNow
            };

            try
            {
                // 1. 파일 준비 완료 대기
                Console.WriteLine($"[ResultCollector] SafePull: Checking file ready...");
                var readyResult = await WaitForFileReadyAsync(deviceSerial, remotePath);

                if (!readyResult.Success)
                {
                    result.Success = false;
                    result.Error = readyResult.Error;
                    return result;
                }

                // 2. 로컬 경로 결정
                if (string.IsNullOrEmpty(localPath))
                {
                    var fileName = Path.GetFileName(remotePath);
                    var subDir = Path.Combine(_localEvidencePath, deviceSerial);

                    if (!Directory.Exists(subDir))
                    {
                        Directory.CreateDirectory(subDir);
                    }

                    localPath = Path.Combine(subDir, fileName);
                }

                result.LocalPath = localPath;

                // 3. ADB Pull 실행 (재시도 로직 포함)
                for (int retry = 0; retry < MAX_RETRY_COUNT; retry++)
                {
                    try
                    {
                        Console.WriteLine($"[ResultCollector] Pulling file (attempt {retry + 1})...");

                        var pullOutput = await ExecuteAdbCommandAsync(
                            deviceSerial,
                            $"pull \"{remotePath}\" \"{localPath}\""
                        );

                        // Pull 성공 확인
                        if (File.Exists(localPath))
                        {
                            var localFileInfo = new FileInfo(localPath);

                            if (localFileInfo.Length >= MIN_FILE_SIZE_BYTES)
                            {
                                result.Success = true;
                                result.FileSize = localFileInfo.Length;
                                result.Duration = DateTime.UtcNow - result.StartTime;

                                Console.WriteLine($"[ResultCollector] Pull SUCCESS: {result.LocalPath} ({result.FileSize} bytes)");
                                return result;
                            }
                        }

                        Console.WriteLine($"[ResultCollector] Pull incomplete, retrying...");
                        await Task.Delay(RETRY_DELAY_MS);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"[ResultCollector] Pull attempt {retry + 1} failed: {ex.Message}");
                        await Task.Delay(RETRY_DELAY_MS);
                    }
                }

                result.Success = false;
                result.Error = "Max retries exceeded";
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Error = ex.Message;
                Console.WriteLine($"[ResultCollector] SafePull error: {ex.Message}");
            }

            result.Duration = DateTime.UtcNow - result.StartTime;
            return result;
        }

        /// <summary>
        /// 여러 파일 일괄 Pull
        /// </summary>
        /// <param name="deviceSerial">기기 시리얼</param>
        /// <param name="remotePaths">원격 파일 경로 목록</param>
        /// <returns>Pull 결과 목록</returns>
        public async Task<BatchPullResult> BatchPullAsync(
            string deviceSerial,
            string[] remotePaths)
        {
            var batchResult = new BatchPullResult
            {
                DeviceSerial = deviceSerial,
                StartTime = DateTime.UtcNow,
                Results = new PullResult[remotePaths.Length]
            };

            for (int i = 0; i < remotePaths.Length; i++)
            {
                batchResult.Results[i] = await SafePullAsync(deviceSerial, remotePaths[i]);
            }

            batchResult.Duration = DateTime.UtcNow - batchResult.StartTime;
            batchResult.SuccessCount = Array.FindAll(batchResult.Results, r => r.Success).Length;
            batchResult.FailCount = remotePaths.Length - batchResult.SuccessCount;

            Console.WriteLine($"[ResultCollector] Batch complete: {batchResult.SuccessCount}/{remotePaths.Length} succeeded");

            return batchResult;
        }

        /// <summary>
        /// 결과 JSON 파일 Pull 및 파싱
        /// </summary>
        /// <param name="deviceSerial">기기 시리얼</param>
        /// <param name="jobId">작업 ID</param>
        /// <returns>결과 데이터</returns>
        public async Task<JobResult> CollectJobResultAsync(string deviceSerial, string jobId)
        {
            var remotePath = $"/sdcard/Scripts/doai-bot/evidence/{jobId}_result.json";

            try
            {
                // 결과 JSON Pull
                var pullResult = await SafePullAsync(deviceSerial, remotePath);

                if (!pullResult.Success)
                {
                    return new JobResult
                    {
                        JobId = jobId,
                        Success = false,
                        Error = pullResult.Error
                    };
                }

                // JSON 파싱
                var content = await File.ReadAllTextAsync(pullResult.LocalPath);
                var jobResult = JsonSerializer.Deserialize<JobResult>(content);

                if (jobResult != null)
                {
                    jobResult.LocalResultPath = pullResult.LocalPath;

                    // 증거 파일 Pull
                    if (jobResult.EvidenceFiles != null && jobResult.EvidenceFiles.Length > 0)
                    {
                        Console.WriteLine($"[ResultCollector] Collecting {jobResult.EvidenceFiles.Length} evidence files...");
                        var evidenceBatch = await BatchPullAsync(deviceSerial, jobResult.EvidenceFiles);
                        jobResult.EvidencePullResults = evidenceBatch.Results;
                    }
                }

                return jobResult ?? new JobResult { JobId = jobId, Success = false, Error = "Parse error" };
            }
            catch (Exception ex)
            {
                return new JobResult
                {
                    JobId = jobId,
                    Success = false,
                    Error = ex.Message
                };
            }
        }

        /// <summary>
        /// ADB 명령 실행
        /// </summary>
        private async Task<string> ExecuteAdbCommandAsync(string deviceSerial, string command)
        {
            var startInfo = new ProcessStartInfo
            {
                FileName = _adbPath,
                Arguments = $"-s {deviceSerial} {command}",
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            using var process = new Process { StartInfo = startInfo };
            process.Start();

            var output = await process.StandardOutput.ReadToEndAsync();
            await process.WaitForExitAsync();

            return output;
        }
    }

    // =============================================
    // 데이터 모델
    // =============================================

    public class ReadySignal
    {
        public string Status { get; set; }
        public string ScreenshotPath { get; set; }
        public long FileSize { get; set; }
        public string Timestamp { get; set; }
        public int CaptureTimeMs { get; set; }
        public int WriteTimeMs { get; set; }
        public string Error { get; set; }
    }

    public class FileReadyResult
    {
        public bool Success { get; set; }
        public string RemotePath { get; set; }
        public long FileSize { get; set; }
        public TimeSpan Duration { get; set; }
        public DateTime StartTime { get; set; }
        public string Method { get; set; }
        public string Error { get; set; }
    }

    public class PullResult
    {
        public bool Success { get; set; }
        public string DeviceSerial { get; set; }
        public string RemotePath { get; set; }
        public string LocalPath { get; set; }
        public long FileSize { get; set; }
        public TimeSpan Duration { get; set; }
        public DateTime StartTime { get; set; }
        public string Error { get; set; }
    }

    public class BatchPullResult
    {
        public string DeviceSerial { get; set; }
        public PullResult[] Results { get; set; }
        public int SuccessCount { get; set; }
        public int FailCount { get; set; }
        public TimeSpan Duration { get; set; }
        public DateTime StartTime { get; set; }
    }

    public class JobResult
    {
        public string JobId { get; set; }
        public bool Success { get; set; }
        public string Timestamp { get; set; }
        public bool VideoFound { get; set; }
        public bool VideoStarted { get; set; }
        public string[] EvidenceFiles { get; set; }
        public int EvidenceCount { get; set; }
        public string Error { get; set; }
        public string LocalResultPath { get; set; }
        public PullResult[] EvidencePullResults { get; set; }
    }
}

File: apps/backend/result-collector.js
/**
 * result-collector.js - Host-side Result Collection Module
 *
 * ADB pull 실행 전 파일 존재 확인 및 동기화 로직
 *
 * Features:
 * 1. ADB shell을 통한 원격 파일 존재 확인
 * 2. .ready 신호 파일 확인 (Android 측 쓰기 완료 신호)
 * 3. 파일 크기 검증 후 pull 실행
 * 4. 재시도 로직 포함
 */

const { execFile } = require('child_process');
const { promisify } = require('util');
const fs = require('fs').promises;
const path = require('path');

const execFileAsync = promisify(execFile);

// =============================================
// 설정 상수
// =============================================
const CONFIG = {
    FILE_CHECK_TIMEOUT_MS: 30000,    // 파일 확인 최대 대기 시간
    FILE_CHECK_INTERVAL_MS: 500,     // 파일 확인 간격
    MIN_FILE_SIZE_BYTES: 1000,       // 최소 파일 크기 (1KB)
    MAX_RETRY_COUNT: 3,              // 최대 재시도 횟수
    RETRY_DELAY_MS: 1000,            // 재시도 간격

    // 기본 경로
    REMOTE_EVIDENCE_PATH: '/sdcard/Scripts/doai-bot/evidence',
    LOCAL_EVIDENCE_PATH: './evidence'
};

// =============================================
// ADB 명령 실행
// =============================================

/**
 * ADB 명령 실행
 * @param {string} adbPath - ADB 실행 파일 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} command - 실행할 명령
 * @returns {Promise<string>} - 명령 출력
 */
async function executeAdbCommand(adbPath, deviceSerial, command) {
    try {
        const args = ['-s', deviceSerial, ...command.split(' ')];
        const { stdout } = await execFileAsync(adbPath, args, {
            timeout: 30000,
            maxBuffer: 10 * 1024 * 1024
        });
        return stdout;
    } catch (error) {
        console.error(`[ResultCollector] ADB error: ${error.message}`);
        throw error;
    }
}

/**
 * ADB shell 명령 실행
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} shellCommand - shell 명령
 * @returns {Promise<string>} - 명령 출력
 */
async function executeAdbShell(adbPath, deviceSerial, shellCommand) {
    return executeAdbCommand(adbPath, deviceSerial, `shell ${shellCommand}`);
}

// =============================================
// 파일 존재/크기 확인
// =============================================

/**
 * 원격 파일 존재 확인
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} remotePath - 원격 파일 경로
 * @returns {Promise<boolean>} - 파일 존재 여부
 */
async function checkRemoteFileExists(adbPath, deviceSerial, remotePath) {
    try {
        const result = await executeAdbShell(
            adbPath,
            deviceSerial,
            `"[ -f '${remotePath}' ] && echo EXISTS || echo NOTFOUND"`
        );
        return result.trim() === 'EXISTS';
    } catch (error) {
        console.error(`[ResultCollector] File check error: ${error.message}`);
        return false;
    }
}

/**
 * 원격 파일 크기 확인
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} remotePath - 원격 파일 경로
 * @returns {Promise<number>} - 파일 크기 (바이트), 실패 시 -1
 */
async function getRemoteFileSize(adbPath, deviceSerial, remotePath) {
    try {
        const result = await executeAdbShell(
            adbPath,
            deviceSerial,
            `stat -c %s '${remotePath}' 2>/dev/null || echo -1`
        );
        const size = parseInt(result.trim(), 10);
        return isNaN(size) ? -1 : size;
    } catch (error) {
        console.error(`[ResultCollector] Size check error: ${error.message}`);
        return -1;
    }
}

// =============================================
// .ready 신호 파일 확인
// =============================================

/**
 * .ready 신호 파일 확인 및 파싱
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} remotePath - 스크린샷 파일 경로
 * @returns {Promise<object>} - 준비 상태 정보
 */
async function checkReadySignal(adbPath, deviceSerial, remotePath) {
    const signalPath = remotePath + '.ready';

    try {
        // 신호 파일 존재 확인
        const exists = await checkRemoteFileExists(adbPath, deviceSerial, signalPath);
        if (!exists) {
            return { status: 'not_found' };
        }

        // 신호 파일 내용 읽기
        const content = await executeAdbShell(
            adbPath,
            deviceSerial,
            `cat '${signalPath}'`
        );

        if (!content || content.trim().length === 0) {
            return { status: 'empty' };
        }

        // JSON 파싱
        try {
            const signal = JSON.parse(content);
            return signal;
        } catch (parseError) {
            return { status: 'parse_error', error: parseError.message };
        }
    } catch (error) {
        return { status: 'error', error: error.message };
    }
}

// =============================================
// 파일 준비 완료 대기 (핵심 로직)
// =============================================

/**
 * 파일 준비 완료까지 대기 (폴링)
 * - .ready 신호 파일 확인 (우선)
 * - 직접 파일 존재/크기 확인 (폴백)
 *
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} remotePath - 원격 파일 경로
 * @param {number} timeoutMs - 타임아웃 (밀리초)
 * @returns {Promise<object>} - { success, fileSize, duration, method, error }
 */
async function waitForFileReady(adbPath, deviceSerial, remotePath, timeoutMs = CONFIG.FILE_CHECK_TIMEOUT_MS) {
    const startTime = Date.now();

    console.log(`[ResultCollector] Waiting for file: ${remotePath}`);

    while (Date.now() - startTime < timeoutMs) {
        // 1. .ready 신호 파일 확인 (우선)
        const signal = await checkReadySignal(adbPath, deviceSerial, remotePath);

        if (signal.status === 'ready') {
            const duration = Date.now() - startTime;
            console.log(`[ResultCollector] File ready (signal): ${signal.file_size} bytes`);
            return {
                success: true,
                fileSize: signal.file_size,
                duration: duration,
                method: 'signal_file'
            };
        }

        if (signal.status === 'failed') {
            console.log(`[ResultCollector] File failed (signal): ${signal.error}`);
            return {
                success: false,
                error: signal.error || 'Remote capture failed',
                duration: Date.now() - startTime,
                method: 'signal_file'
            };
        }

        // 2. 직접 파일 확인 (신호 파일 없는 경우)
        const exists = await checkRemoteFileExists(adbPath, deviceSerial, remotePath);

        if (exists) {
            const size = await getRemoteFileSize(adbPath, deviceSerial, remotePath);

            if (size >= CONFIG.MIN_FILE_SIZE_BYTES) {
                // 크기 안정화 확인 (1초 후 재확인)
                await sleep(1000);
                const size2 = await getRemoteFileSize(adbPath, deviceSerial, remotePath);

                if (size2 === size) {
                    const duration = Date.now() - startTime;
                    console.log(`[ResultCollector] File ready (direct): ${size} bytes`);
                    return {
                        success: true,
                        fileSize: size,
                        duration: duration,
                        method: 'direct_check'
                    };
                }
            }
        }

        await sleep(CONFIG.FILE_CHECK_INTERVAL_MS);
    }

    // 타임아웃
    console.log(`[ResultCollector] File timeout: ${remotePath}`);
    return {
        success: false,
        error: 'Timeout waiting for file',
        duration: Date.now() - startTime
    };
}

// =============================================
// 안전한 ADB Pull
// =============================================

/**
 * 안전한 ADB Pull (파일 준비 확인 후 실행)
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} remotePath - 원격 파일 경로
 * @param {string} localPath - 로컬 저장 경로 (null이면 자동 생성)
 * @returns {Promise<object>} - Pull 결과
 */
async function safePull(adbPath, deviceSerial, remotePath, localPath = null) {
    const result = {
        deviceSerial,
        remotePath,
        startTime: new Date().toISOString()
    };

    try {
        // 1. 파일 준비 완료 대기
        console.log(`[ResultCollector] SafePull: Checking file ready...`);
        const readyResult = await waitForFileReady(adbPath, deviceSerial, remotePath);

        if (!readyResult.success) {
            result.success = false;
            result.error = readyResult.error;
            return result;
        }

        // 2. 로컬 경로 결정
        if (!localPath) {
            const fileName = path.basename(remotePath);
            const subDir = path.join(CONFIG.LOCAL_EVIDENCE_PATH, deviceSerial);

            await fs.mkdir(subDir, { recursive: true });
            localPath = path.join(subDir, fileName);
        }

        result.localPath = localPath;

        // 3. ADB Pull 실행 (재시도 로직)
        for (let retry = 0; retry < CONFIG.MAX_RETRY_COUNT; retry++) {
            try {
                console.log(`[ResultCollector] Pulling file (attempt ${retry + 1})...`);

                await executeAdbCommand(
                    adbPath,
                    deviceSerial,
                    `pull "${remotePath}" "${localPath}"`
                );

                // Pull 성공 확인
                try {
                    const stats = await fs.stat(localPath);

                    if (stats.size >= CONFIG.MIN_FILE_SIZE_BYTES) {
                        result.success = true;
                        result.fileSize = stats.size;
                        result.endTime = new Date().toISOString();

                        console.log(`[ResultCollector] Pull SUCCESS: ${localPath} (${stats.size} bytes)`);
                        return result;
                    }
                } catch (statError) {
                    // 파일 없음
                }

                console.log(`[ResultCollector] Pull incomplete, retrying...`);
                await sleep(CONFIG.RETRY_DELAY_MS);
            } catch (pullError) {
                console.log(`[ResultCollector] Pull attempt ${retry + 1} failed: ${pullError.message}`);
                await sleep(CONFIG.RETRY_DELAY_MS);
            }
        }

        result.success = false;
        result.error = 'Max retries exceeded';
    } catch (error) {
        result.success = false;
        result.error = error.message;
        console.error(`[ResultCollector] SafePull error: ${error.message}`);
    }

    result.endTime = new Date().toISOString();
    return result;
}

/**
 * 여러 파일 일괄 Pull
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string[]} remotePaths - 원격 파일 경로 목록
 * @returns {Promise<object>} - Batch Pull 결과
 */
async function batchPull(adbPath, deviceSerial, remotePaths) {
    const batchResult = {
        deviceSerial,
        startTime: new Date().toISOString(),
        results: []
    };

    for (const remotePath of remotePaths) {
        const result = await safePull(adbPath, deviceSerial, remotePath);
        batchResult.results.push(result);
    }

    batchResult.endTime = new Date().toISOString();
    batchResult.successCount = batchResult.results.filter(r => r.success).length;
    batchResult.failCount = remotePaths.length - batchResult.successCount;

    console.log(`[ResultCollector] Batch complete: ${batchResult.successCount}/${remotePaths.length} succeeded`);

    return batchResult;
}

// =============================================
// Job 결과 수집 (통합)
// =============================================

/**
 * Job 결과 JSON 파일 Pull 및 파싱
 * @param {string} adbPath - ADB 경로
 * @param {string} deviceSerial - 기기 시리얼
 * @param {string} jobId - 작업 ID
 * @returns {Promise<object>} - Job 결과
 */
async function collectJobResult(adbPath, deviceSerial, jobId) {
    const remotePath = `${CONFIG.REMOTE_EVIDENCE_PATH}/${jobId}_result.json`;

    try {
        // 1. 결과 JSON Pull
        const pullResult = await safePull(adbPath, deviceSerial, remotePath);

        if (!pullResult.success) {
            return {
                jobId,
                success: false,
                error: pullResult.error
            };
        }

        // 2. JSON 파싱
        const content = await fs.readFile(pullResult.localPath, 'utf-8');
        const jobResult = JSON.parse(content);

        jobResult.localResultPath = pullResult.localPath;

        // 3. 증거 파일 Pull
        if (jobResult.evidence_files && jobResult.evidence_files.length > 0) {
            console.log(`[ResultCollector] Collecting ${jobResult.evidence_files.length} evidence files...`);
            const evidenceBatch = await batchPull(adbPath, deviceSerial, jobResult.evidence_files);
            jobResult.evidencePullResults = evidenceBatch.results;
        }

        return jobResult;
    } catch (error) {
        return {
            jobId,
            success: false,
            error: error.message
        };
    }
}

// =============================================
// 유틸리티
// =============================================

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// =============================================
// 모듈 Export
// =============================================
module.exports = {
    // 파일 확인
    checkRemoteFileExists,
    getRemoteFileSize,
    checkReadySignal,

    // 파일 대기 (핵심)
    waitForFileReady,

    // ADB Pull
    safePull,
    batchPull,

    // Job 결과 수집
    collectJobResult,

    // ADB 명령
    executeAdbCommand,
    executeAdbShell,

    // 설정
    CONFIG
};
